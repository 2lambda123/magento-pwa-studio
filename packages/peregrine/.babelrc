{
  "plugins": [
    "transform-object-rest-spread",
    "transform-class-properties",
    ["transform-react-jsx", { "pragma": "createElement" }],
    ["transform-runtime", {
      "helpers": true,
      "polyfill": false, // polyfills will be handled by preset-env
      "regenerator": false
    }]
  ],
  "presets": [
    ["env", {
      "targets": {
        "browsers": ["> 5%"]
      },
      "modules": false
    }]
  ],
  "env": {
    "test": {
      "plugins": [
        "transform-es2015-modules-commonjs",
        "transform-object-rest-spread",
        "transform-class-properties",
        ["transform-react-jsx", { "pragma": "createElement" }]
      ]
    }
  }
}
