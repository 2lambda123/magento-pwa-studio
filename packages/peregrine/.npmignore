.eslintrc.js
.storybook
/lib/**/__{docs,helpers,mocks,stories,tests}__/**
/scripts
