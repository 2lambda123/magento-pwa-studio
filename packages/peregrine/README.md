# Peregrine

Peregrine is a collection of UI components for Magento PWA projects.

## Under construction

This repository is currently under reconstruction and is not suitable for use outside of the Venia storefront. Stay tuned for more information on the vision for `peregrine`.

## Contributing

The `peregrine` repository is an open source project that welcomes contributors of all skill levels.

If you want to contribute to this project, please review the [contribution guidelines] and follow our [code of conduct].

## License

This project is under the OSL-3.0 license - see the [LICENSE] file for details.

[Peregrine documentation]: https://magento-research.github.io/pwa-studio/peregrine/
[contribution guidelines]: .github/CONTRIBUTING.md
[code of conduct]: .github/CODE_OF_CONDUCT.md
[LICENSE]: LICENSE
