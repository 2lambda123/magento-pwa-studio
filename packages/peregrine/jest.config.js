module.exports = {
    displayName: 'Peregrine',
    setupFiles: [
        '<rootDir>/scripts/shim.js',
        '<rootDir>/scripts/fetch-mock.js'
    ],
    browser: true,
    testURL: 'https://localhost/',
    coveragePathIgnorePatterns: ['scripts/*', 'node_modules', 'src/index.js'],
    testPathIgnorePatterns: ['dist', 'node_modules']
};
