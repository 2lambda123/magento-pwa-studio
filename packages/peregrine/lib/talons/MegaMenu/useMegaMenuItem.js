import { useCallback, useMemo, useState } from 'react';

export const useMegaMenuItem = props => {
    const { category, activeCategoryId, subMenuState, disableFocus } = props;

    const [isFocused, setIsFocused] = useState(false);
    const isActive = category.id === activeCategoryId;

    const handleCloseSubMenu = useCallback(() => {
        setIsFocused(false);
    }, [setIsFocused]);

    const isMenuActive = useMemo(() => {
        if (!isFocused) {
            return false;
        }
        if (subMenuState) {
            return true;
        } else if (disableFocus) {
            setIsFocused(false);
        }
        return false;
    }, [isFocused, subMenuState, disableFocus]);

    const a11yClick = e => {
        //checking down arrow
        if (e.key === 'ArrowDown' || e.key === ' ') {
            return true;
        }
        //checking up arrow or escape
        if (e.key === 'ArrowUp' || e.key === 'Escape') {
            setIsFocused(false);
        }
        //checking Tab with Shift
        if (e.shiftKey && e.key === 'Tab') {
            setIsFocused(false);
        }
    };

    const toggleSubMenu = e => {
        e.preventDefault();
        if (
            category.children.length &&
            !(e.key === 'ArrowUp' || e.key === 'Escape')
        ) {
            setIsFocused(true);
        } else {
            setIsFocused(false);
        }
    };

    const handleKeyDown = e => {
        a11yClick(e) && toggleSubMenu(e);
    };

    return {
        isFocused,
        isActive,
        handleCloseSubMenu,
        isMenuActive,
        handleKeyDown
    };
};
