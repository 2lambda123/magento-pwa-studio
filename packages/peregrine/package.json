{
  "name": "@magento/peregrine",
  "version": "3.0.0",
  "publishConfig": {
    "access": "public"
  },
  "description": "The core runtime of Magento PWA",
  "main": "lib/index.js",
  "scripts": {
    "clean": " ",
    "storybook": "start-storybook -p 9001 -c .storybook",
    "storybook:build": "build-storybook -c .storybook -o storybook-dist"
  },
  "repository": "github:magento-research/pwa-studio",
  "author": "Magento Commerce",
  "license": "(OSL-3.0 OR AFL-3.0)",
  "bugs": {
    "url": "https://github.com/magento-research/pwa-studio/issues"
  },
  "homepage": "https://github.com/magento-research/pwa-studio/tree/master/packages/peregrine#readme",
  "devDependencies": {
    "@storybook/react": "~4.1.13",
    "intl": "~1.2.5",
    "react": "~16.8.3",
    "react-apollo": "~2.5.3",
    "react-dom": "~16.8.3",
    "react-router-dom": "~5.0.0",
    "react-test-renderer": "~16.8.6"
  },
  "peerDependencies": {
    "@babel/runtime": "^7.4.2",
    "react": "^16.8.0",
    "react-apollo": "^2.5.3",
    "react-dom": "^16.8.0",
    "react-router-dom": "~5.0.0"
  },
  "engines": {
    "node": "10.x",
    "yarn": ">=1.12.0"
  },
  "module": "lib/index.js",
  "jsnext:main": "lib/index.js",
  "es2015": "lib/index.js",
  "sideEffects": false
}
