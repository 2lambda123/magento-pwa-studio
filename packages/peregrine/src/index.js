import * as RestApi from './RestApi';

export { default } from './Peregrine';
export { default as ContainerChild } from './ContainerChild';
export { default as List, Items, Item } from './List';
export { default as Router } from './Router';
export { default as Simulators } from './Simulators';
export { default as Price } from './Price';
export { RestApi };
