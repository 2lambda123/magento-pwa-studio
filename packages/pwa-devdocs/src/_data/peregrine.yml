title: Peregrine
entries:
  - label: Overview
    url: /peregrine/

  - label: Storybook UI testing
    url: /peregrine/testing/storybook/

  - label: Utility modules
    entries:

      - label: REST API client
        url: /peregrine/reference/rest-api-client/

  - label: React components
    entries:

      - label: ContainerChild
        url: /peregrine/reference/container-child/

      - label: Item
        url: /peregrine/reference/item/

      - label: Items
        url: /peregrine/reference/items/

      - label: List
        url: /peregrine/reference/list/

      - label: Router
        url: /peregrine/reference/router/
