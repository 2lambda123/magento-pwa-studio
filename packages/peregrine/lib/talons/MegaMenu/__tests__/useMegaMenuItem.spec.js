import React from 'react';

import { useMegaMenuItem } from '../useMegaMenuItem';
import { createTestInstance } from '@magento/peregrine';

const Component = props => {
    const talonProps = useMegaMenuItem(props);

    return <i talonProps={talonProps} />;
};

const getTalonProps = props => {
    const tree = createTestInstance(<Component {...props} />);
    const { root } = tree;
    const { talonProps } = root.findByType('i').props;

    const update = newProps => {
        tree.update(<Component {...{ ...props, ...newProps }} />);

        return root.findByType('i').props.talonProps;
    };

    return { talonProps, tree, update };
};

const defaultProps = {
    category: {
        id: 1,
        name: 'Women',
        url_path: 'women',
        url_suffix: '.html',
        isActive: true,
        children: [
            {
                id: 2,
                name: 'Bottoms',
                url_path: 'bottoms',
                url_suffix: '.html',
                children: []
            },
            {
                id: 3,
                name: 'Tops',
                url_path: 'tops',
                url_suffix: '.html',
                isActive: false,
                children: []
            }
        ]
    },
    activeCategoryId: 1,
    subMenuState: true
};

jest.mock('react', () => {
    const React = jest.requireActual('react');
    const memoSpy = jest.spyOn(React, 'useMemo');
    const stateSpy = jest.spyOn(React, 'useState');
    const callbackSpy = jest.spyOn(React, 'useCallback');

    return Object.assign(React, {
        useMemo: memoSpy,
        useState: stateSpy,
        useCallback: callbackSpy
    });
});

test('Should return correct shape', () => {
    const { talonProps } = getTalonProps(defaultProps);

    expect(talonProps).toMatchSnapshot();
});

test('handleCloseSubMenu to be called', () => {
    const props = {
        ...defaultProps
    };
    const { talonProps } = getTalonProps(props);

    talonProps.handleCloseSubMenu();

    expect(talonProps.isMenuActive).toEqual(false);
});

test('a11yClick to be called', () => {
    const e = { keyCode: '27' };

    const props = {
        ...defaultProps
    };
    const { talonProps } = getTalonProps(props);

    talonProps.a11yClick(e);

    expect(talonProps.isFocused).toEqual(false);
});

test('toggleSubMenu to be called', () => {
    const e = { keyCode: '27', preventDefault: jest.fn() };
    const props = {
        ...defaultProps
    };
    const { talonProps } = getTalonProps(props);

    talonProps.toggleSubMenu(e);

    expect(talonProps.isActive).toEqual(true);
});
