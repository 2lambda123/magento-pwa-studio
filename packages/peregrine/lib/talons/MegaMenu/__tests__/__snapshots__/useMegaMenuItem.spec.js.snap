// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should return correct shape 1`] = `
Object {
  "a11yClick": [Function],
  "handleCloseSubMenu": [Function],
  "isActive": true,
  "isFocused": true,
  "isMenuActive": true,
  "toggleSubMenu": [Function],
}
`;
